Source/WebKit/NetworkProcess/EarlyHintsResourceLoader.cpp

```diff
@@ -0,0 +1,141 @@
+        return;
+
+    // Only the first early hint response served during the navigation is handled.
+    // FIXME: discard hints on cross-origin redirect once we support early hint preloads.
+    if (m_hasReceivedEarlyHints)
+        return;
+    m_hasReceivedEarlyHints = true;
+
+    auto headerValue = response.httpHeaderField(HTTPHeaderName::Link);
+    if (headerValue.isEmpty())
+        return;
+
+    auto url = response.url();
+    ContentSecurityPolicy contentSecurityPolicy { URL { url }, &m_loader, nullptr };
+    contentSecurityPolicy.didReceiveHeaders(ContentSecurityPolicyResponseHeaders { response }, originalRequest().httpReferrer());
+
+    LinkHeaderSet headerSet(headerValue);
+    for (const auto& header : headerSet) {
+        if (!header.valid() || header.url().isEmpty() || header.rel().isEmpty() || header.isViewportDependent())
+            continue;
+
+        if (equalLettersIgnoringASCIICase(header.rel(), "preconnect"_s))
+            startPreconnectTask(response.url(), header, contentSecurityPolicy);
+    }
+}
+
+ResourceRequest EarlyHintsResourceLoader::constructPreconnectRequest(const URL& url)
+{
+    ResourceRequest request { url };
+
+    // firstPartyForCookies and user agent are part of the HTTP socket pool keys in CFNetwork: rdar://59434166
+    auto firstPartyForCookies = originalRequest().firstPartyForCookies();
+    if (firstPartyForCookies.isValid())
+        request.setFirstPartyForCookies(firstPartyForCookies);
+
+    auto userAgent = originalRequest().httpUserAgent();
+    if (!userAgent.isEmpty())
+        request.setHTTPUserAgent(userAgent);
+
+    return request;
+}
+
+void EarlyHintsResourceLoader::startPreconnectTask(const URL& baseURL, const LinkHeader& header, const ContentSecurityPolicy& contentSecurityPolicy)
+{
+    if (!m_loader.parameters().linkPreconnectEarlyHintsEnabled)
+        return;
+
+    URL url(baseURL, header.url());
+    if (!url.isValid() || !url.protocolIsInHTTPFamily())
+        return;
+
+    if (!contentSecurityPolicy.allowConnectToSource(url, ContentSecurityPolicy::RedirectResponseReceived::No, originalRequest().url()))
+        return;
+
+    auto* networkSession = m_loader.connectionToWebProcess().networkSession();
+    if (!networkSession)
+        return;
+
+    NetworkLoadParameters parameters;
+    auto globalFrameID = m_loader.globalFrameID();
+    parameters.webPageProxyID = globalFrameID.webPageProxyID;
+    parameters.webPageID = globalFrameID.webPageID;
+    parameters.webFrameID = globalFrameID.frameID;
+    parameters.storedCredentialsPolicy = equalLettersIgnoringASCIICase(header.crossOrigin(), "anonymous"_s) ? StoredCredentialsPolicy::DoNotUse : StoredCredentialsPolicy::Use;
+    parameters.contentSniffingPolicy = ContentSniffingPolicy::DoNotSniffContent;
+    parameters.contentEncodingSniffingPolicy = ContentEncodingSniffingPolicy::Default;
```

Comment:
Why do we need these two?

Reply:
Almost certainly unnecessary but I was copying what we do in `SpeculativeLoadManager::preconnectForSubresource`. I'll remove it.